File: Recursion/Fibonacci Number/By iterative.cpp
// Method - Iterative(Brute Force)
// Time Complexity T - O(n), since loop run for "n"
// Space Complexity S - O(1), Since brute force don't take extra space


class Solution {
public:
    int fib(int n) {
        // if the value of n is 1 or less than 1
        // then return "n"
        if (n <= 1){
            return n;
        }
        
        
        // if the value of "n" is 2
        // then return "1"
        if (n == 2){
            return 1;
        }
        
        // setting the current to "0", prev1 to "1" and prev2 to "1"
        int current = 0;
        int prev1 = 1;
        int prev2 = 1;
        
        
        // loop through 3 to "n"
        for(int i = 3; i<=n ; i++){
            
            // adding the prev1 and prev2 value for calculating the factorial
            current = prev1+prev2;
            prev2 = prev1;
            prev1 = current;
        }
        return current;
    }
};

File: Array/Majority Element/By iterative.java
// Method - Iterative(BruteForce)
// Time Complexity T - O(n^2), since nested loop runs n^2
// Space Complexity S - O(1), brute force does't take extra sapce

class Solution {
    public int majorityElement(int[] nums) {
        
        // the count of majority element in the given array is greater than n/2
        int majorityElementCount = nums.length/2; 
        
        // loop runs for all the values in nums
        for(int num:nums){
            
            // elementCount is used to count the occurance of an element
            int elementCount = 0;
            
            // loop runs for all the values in nums
            for (int element : nums){
                
                // if i == j, 
                // then increment the count of an element
                if(element == num){
                    elementCount = elementCount + 1;
                }
            }
            
            // the occurence of an element is greater than the majorityElementCount
            // then print the element(number)
            if (elementCount > majorityElementCount){
                return num;
            }
        }
        
        // if there is no element occured more then the majority element count
        // then return -1
        return -1;
    }
}
